virtual_ship/sailship.py

```diff
@@ -3,185 +3,107 @@
     argo_floats: list[ArgoFloat] = []
-
-    # drifter deployment locations that have been visited
-    drifter_locations_visited: set[Location] = set()
-    # drifter objects to be used in simulation
     drifters: list[Drifter] = []
-
-    # ctd cast locations that have been visited
-    ctd_locations_visited: set[Location] = set()
-    # ctd cast objects to be used in simulation
     ctds: list[CTD] = []
 
-    # iterate over each discrete route point, find deployment and measurement locations and times, and measure how much time this took
-    # TODO between drifters, argo floats, ctd there is quite some repetition
-    print("Traversing ship route.")
-    time_past = timedelta()
-    for i, route_point in enumerate(route_points):
-        if i % 96 == 0:
-            print(f"Gathered data {time_past} hours since start.")
-
-        # find drifter deployments to be done at this location
-        drifters_here = set(
-            [
-                drifter
-                for drifter in drifter_locations - drifter_locations_visited
-                if all(
-                    np.isclose(
-                        [drifter.lat, drifter.lon], [route_point.lat, route_point.lon]
+    assert (
+        config.waypoints[0].time is not None
+    )  # cannot happen as verify_waypoints checks this
+    time = config.waypoints[0].time
+
+    for wp, wp_next in zip(
+        config.waypoints, config.waypoints[1:] + [None], strict=True
+    ):
+        spacetime = Spacetime(location=wp.location, time=time)
+
+        match wp.instrument:
+            case None:
+                pass
+            case InstrumentType.CTD:
+                ctds.append(
+                    CTD(
+                        spacetime=spacetime,
+                        min_depth=config.ctd_fieldset.U.depth[0],
+                        max_depth=config.ctd_fieldset.U.depth[-1],
                     )
                 )
-            ]
-        )
-        if len(drifters_here) > 1:
-            print(
-                "WARN: Multiple drifter deployments match the current location. Only a single deployment will be performed."
-            )
-        drifters.append(
-            Drifter(
-                spacetime=Spacetime(
-                    location=route_point, time=time_past.total_seconds()
-                ),
-                depth=-config.drifter_fieldset.U.depth[0],
-                lifetime=timedelta(weeks=4),
-            )
-        )
-        drifter_locations_visited = drifter_locations_visited.union(drifters_here)
-
-        # find argo float deployments to be done at this location
-        argos_here = set(
-            [
-                argo
-                for argo in argo_locations - argo_locations_visited
-                if all(
-                    np.isclose([argo.lat, argo.lon], [route_point.lat, route_point.lon])
+                # use 20 minutes to case the CTD
+                time += timedelta(minutes=20)
```

Comment:
I hadn't seen this before. Does it make sense to have this hardcoded? Can confuse users who think their planning fits, but who don't realise another 20 minutes is added for each CTD. Perhaps add a warning/notice? @ammedd, what do you think?

Reply:
it is a parameter in the config now